eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/PlaceSuggestionModule.java

```diff
@@ -0,0 +1,202 @@
+             */
+
+            googleApiClient.connect();
+            return new PlaceSuggestionModule(changeableList, itemSelectionAction, googleApiClient);
+        }
+
+    };
+
+    private static final String TAG = PlaceSuggestionModule.class.getSimpleName();
+
+    private final NonNotifyingChangeableList<ListItem> mChangeableList;
+
+    private ListItemSelectionAction<GeoPlace> mSelectionAction;
+
+    private GoogleApiClient mGoogleApiClient;
+    private ExecutorService mExecutorService;
+
+    private PlaceSuggestionQueryTask.Client mSuggestionQueryTaskClient;
+    private String mLastQuery;
+
+    private PlaceByIdTaskClient mPlaceByIdTaskClient;
+
+
+    private PlaceSuggestionModule(NonNotifyingChangeableList<ListItem> changeableList,
+                                  ListItemSelectionAction<GeoPlace> itemSelectionAction,
+                                  GoogleApiClient googleApiClient)
+    {
+        mChangeableList = changeableList;
+        mSelectionAction = itemSelectionAction;
+        mGoogleApiClient = googleApiClient;
+
+        mExecutorService = Executors.newSingleThreadExecutor();
+        mSuggestionQueryTaskClient = new PlaceSuggestionQueryTaskClient();
+        mPlaceByIdTaskClient = new PlaceByIdTaskClient();
+    }
+
+
+    @Override
+    public void shutDown()
+    {
+        mGoogleApiClient.disconnect();
+
+        // TODO think about this, is callback better? That would have to be nulled out as well.
+        mSelectionAction = null;
+    }
+
+
+    @Override
+    public void onPlaceSuggestionSelected(NamedPlace namedPlace)
+    {
+        new PlaceByIdTask(namedPlace, mPlaceByIdTaskClient).executeOnExecutor(mExecutorService, mGoogleApiClient);
+    }
+
+
+    @Override
+    public void onSearchQueryChange(String newQuery)
+    {
+        mLastQuery = newQuery;
+        new PlaceSuggestionQueryTask(newQuery, mSuggestionQueryTaskClient).executeOnExecutor(mExecutorService, mGoogleApiClient);
+    }
+
+
+    private void onSuggestionsReceived(List<ListItem> newSuggestions)
+    {
+        if (newSuggestions.isEmpty())
+        {
+            mChangeableList.apply(new ClearAll<>());
+            return;
+        }
+
+        for (ListItem newItem : newSuggestions)
+        {
+            if (newItem instanceof PlaceSuggestionItem)
+            {
+                ((PlaceSuggestionItem) newItem).setListener(PlaceSuggestionModule.this);
+            }
+        }
+
+        // TODO update with real header
+        if (!newSuggestions.isEmpty())
+        {
+            newSuggestions.add(0, new DummyHeaderItem("Suggestions"));
+        }
+
+        mChangeableList.apply(new ReplaceAll<>(newSuggestions));
+    }
+
+
+    private final class PlaceSuggestionQueryTaskClient implements PlaceSuggestionQueryTask.Client
+    {
+
+        @Override
+        public boolean shouldDiscard(String query)
+        {
+            return !Objects.equals(query, mLastQuery);
```

Comment:
IMO this shouldn't be handled here. In particular `mLastQuery` looks like an inappropriate piece of mutable state.

As suggested on slack I'd prefer a different control flow. Which is, a `SearchResult` is introduced that holds the actual results as well as the query that it was launched for. The `SearchResult` is then returned to the owner of the module. When applying the search result, the query will be checked against the current query.

To cancel any pending tasks you should take control over the execution queue:

```java
mJobQueue = new LinkedBlockingQueue<Runnable>();
mExecutorService = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, mJobQueue)
```

Whenever a new query arrives you just wipe the queue to cancel pending requests.
```java
mJobQueue.clear();
```